File: 54. Spiral Matrix/Solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        for (int top = 0, left = 0, bottom = matrix.length - 1, right = matrix[0].length - 1; top <= bottom && left <= right; top++, bottom--, left++, right--) {
            for (int i = left; i <= right; i++) res.add(matrix[top][i]);
            if (top == bottom) break;
            for (int i = top + 1; i <= bottom; i++) res.add(matrix[i][right]);
            if (left == right) break;
            for (int i = right - 1; i >= left; i--) res.add(matrix[bottom][i]);
            for (int i = bottom - 1; i >= top + 1; i--) res.add(matrix[i][left]);
        }
        return res;
    }
}

File: 54. Spiral Matrix/Solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        top, left, bottom, right = 0, 0, len(matrix) - 1, len(matrix[0]) - 1
        res = []
        while top <= bottom and left <= right:
            for i in range(left, right + 1): res.append(matrix[top][i])
            if top == bottom: break
            for i in range(top + 1, bottom + 1): res.append(matrix[i][right])
            if left == right: break
            for i in range(right - 1, left - 1, -1): res.append(matrix[bottom][i])
            for i in range(bottom - 1, top, -1): res.append(matrix[i][left])
            top += 1
            bottom -= 1
            left += 1
            right -= 1
        return res